File: bridge/src/wsjtx.ts
/**
 * ProPulse Bridge — WSJT-X UDP Protocol Parser
 *
 * Listens for WSJT-X UDP datagrams and parses the QDataStream binary
 * format. Extracts Status, Decode, Clear, and QSO Logged messages.
 *
 * Protocol reference:
 *   https://sourceforge.net/p/wsjt/wsjtx/ci/master/tree/Network/NetworkMessage.hpp
 *
 * QDataStream encoding (big-endian):
 *   - uint32:  4 bytes
 *   - utf8:    uint32 length + bytes  (0xFFFFFFFF = null)
 *   - bool:    uint8
 *   - double:  8 bytes IEEE 754
 *   - QTime:   uint32 milliseconds since midnight
 */

import { createSocket, type Socket as UDPSocket } from "node:dgram";
import type { WSJTXStatus, WSJTXDecode, WSJTXQSOLogged } from "./types.js";

// ============================================================================
// Constants
// ============================================================================

const WSJTX_MAGIC = 0xadbccbda;
const WSJTX_SCHEMA_VERSION = 2;

/** WSJT-X network message types */
const enum WSJTXMessageType {
  Heartbeat = 0,
  Status = 1,
  Decode = 2,
  Clear = 3,
  // Reply = 4,  (outbound only)
  QSOLogged = 5,
  // Close = 6,
  // Replay = 7,
  // HaltTx = 8,
  // FreeText = 9,
  // WSPRDecode = 10,
  // Location = 11,
  LoggedADIF = 12,
}

// ============================================================================
// QDataStream Reader
// ============================================================================

class QDataStreamReader {
  private offset = 0;

  constructor(private readonly buf: Buffer) {}

  get remaining(): number {
    return this.buf.length - this.offset;
  }

  readUInt8(): number {
    if (this.remaining < 1) throw new RangeError("Buffer underflow: uint8");
    const v = this.buf.readUInt8(this.offset);
    this.offset += 1;
    return v;
  }

  readUInt32(): number {
    if (this.remaining < 4) throw new RangeError("Buffer underflow: uint32");
    const v = this.buf.readUInt32BE(this.offset);
    this.offset += 4;
    return v;
  }

  readInt32(): number {
    if (this.remaining < 4) throw new RangeError("Buffer underflow: int32");
    const v = this.buf.readInt32BE(this.offset);
    this.offset += 4;
    return v;
  }

  readBool(): boolean {
    return this.readUInt8() !== 0;
  }

  readDouble(): number {
    if (this.remaining < 8) throw new RangeError("Buffer underflow: double");
    const v = this.buf.readDoubleBE(this.offset);
    this.offset += 8;
    return v;
  }

  readQTime(): number {
    return this.readUInt32(); // ms since midnight
  }

  /**
   * Read a QDataStream UTF-8 string.
   * Length prefix of 0xFFFFFFFF means null/empty.
   */
  readUtf8(): string | null {
    const len = this.readUInt32();
    if (len === 0xffffffff) return null;
    if (this.remaining < len)
      throw new RangeError("Buffer underflow: utf8 string");
    const s = this.buf.toString("utf-8", this.offset, this.offset + len);
    this.offset += len;
    return s;
  }

  /**
   * Skip forward by `n` bytes.
   */
  skip(n: number): void {
    if (this.remaining < n) throw new RangeError("Buffer underflow: skip");
    this.offset += n;
  }
}

// ============================================================================
// FT8/FT4 Message Text Parsing
// ============================================================================

/** Regex: CQ CALLSIGN GRID */
const CQ_REGEX =
  /^CQ\s+(?:(\w{2,4})\s+)?([A-Z0-9/]{3,})\s+([A-R]{2}\d{2}(?:[a-x]{2})?)$/i;

/** Regex: CALL1 CALL2 GRID */
const QSO_GRID_REGEX =
  /^([A-Z0-9/]{3,})\s+([A-Z0-9/]{3,})\s+([A-R]{2}\d{2}(?:[a-x]{2})?)$/i;

/** Regex: CALL1 CALL2 REPORT  (e.g. +05 or -15 or R+05 or R-15 or RR73 or 73) */
const QSO_REPORT_REGEX =
  /^([A-Z0-9/]{3,})\s+([A-Z0-9/]{3,})\s+(R?[+-]\d{2}|RR73|RRR|73)$/i;

interface ExtractedCall {
  callsign?: string;
  grid?: string;
}

function extractCallInfo(message: string): ExtractedCall {
  const trimmed = message.trim();

  // CQ [directed] CALL GRID
  const cqMatch = trimmed.match(CQ_REGEX);
  if (cqMatch) {
    return { callsign: cqMatch[2], grid: cqMatch[3] };
  }

  // CALL1 CALL2 GRID
  const gridMatch = trimmed.match(QSO_GRID_REGEX);
  if (gridMatch) {
    return { callsign: gridMatch[2], grid: gridMatch[3] };
  }

  // CALL1 CALL2 REPORT
  const reportMatch = trimmed.match(QSO_REPORT_REGEX);
  if (reportMatch) {
    return { callsign: reportMatch[2] };
  }

  return {};
}

// ============================================================================
// Decode Statistics Tracker
// ============================================================================

interface DecodeStats {
  recentDecodes: number[];
  uniqueCallsigns: Map<string, number>;
}

const STATS_WINDOW_MS = 15 * 60 * 1000; // 15 minutes

function createDecodeStats(): DecodeStats {
  return {
    recentDecodes: [],
    uniqueCallsigns: new Map(),
  };
}

function recordDecode(stats: DecodeStats, callsign?: string): void {
  const now = Date.now();
  stats.recentDecodes.push(now);

  if (callsign) {
    stats.uniqueCallsigns.set(callsign, now);
  }

  // Prune old entries
  const cutoff = now - STATS_WINDOW_MS;
  stats.recentDecodes = stats.recentDecodes.filter((t) => t > cutoff);
  for (const [call, ts] of stats.uniqueCallsigns) {
    if (ts < cutoff) stats.uniqueCallsigns.delete(call);
  }
}

// ============================================================================
// Event Handler Types
// ============================================================================

type StatusHandler = (status: WSJTXStatus, instanceId: string) => void;
type DecodeHandler = (decode: WSJTXDecode, instanceId: string) => void;
type QSOLoggedHandler = (qso: WSJTXQSOLogged, instanceId: string) => void;
type ClearHandler = (window: string | undefined, instanceId: string) => void;
type ErrorHandler = (error: Error) => void;

// ============================================================================
// WSJTXListener
// ============================================================================

export class WSJTXListener {
  private socket: UDPSocket | null = null;
  private readonly port: number;
  private readonly statsPerInstance = new Map<string, DecodeStats>();

  // Event handlers
  private statusHandlers: StatusHandler[] = [];
  private decodeHandlers: DecodeHandler[] = [];
  private qsoLoggedHandlers: QSOLoggedHandler[] = [];
  private clearHandlers: ClearHandler[] = [];
  private errorHandlers: ErrorHandler[] = [];

  constructor(port = 2237) {
    this.port = port;
  }

  // --------------------------------------------------------------------------
  // Event Registration
  // --------------------------------------------------------------------------

  onStatus(handler: StatusHandler): void {
    this.statusHandlers.push(handler);
  }

  onDecode(handler: DecodeHandler): void {
    this.decodeHandlers.push(handler);
  }

  onQSOLogged(handler: QSOLoggedHandler): void {
    this.qsoLoggedHandlers.push(handler);
  }

  onClear(handler: ClearHandler): void {
    this.clearHandlers.push(handler);
  }

  onError(handler: ErrorHandler): void {
    this.errorHandlers.push(handler);
  }

  // --------------------------------------------------------------------------
  // Lifecycle
  // --------------------------------------------------------------------------

  start(): void {
    if (this.socket) return;

    this.socket = createSocket("udp4");

    this.socket.on("message", (msg: Buffer) => {
      try {
        this.handleDatagram(msg);
      } catch (err) {
        this.emitError(err instanceof Error ? err : new Error(String(err)));
      }
    });

    this.socket.on("error", (err: Error) => {
      this.emitError(err);
    });

    this.socket.bind(this.port, "127.0.0.1");
  }

  stop(): void {
    if (this.socket) {
      this.socket.close();
      this.socket = null;
    }
    this.statsPerInstance.clear();
  }

  /** Returns decode statistics for all known instances */
  getStats(): Record<
    string,
    { decodesPerMinute: number; uniqueCallsigns: number }
  > {
    const result: Record<
      string,
      { decodesPerMinute: number; uniqueCallsigns: number }
    > = {};
    const now = Date.now();
    const oneMinuteAgo = now - 60_000;

    for (const [id, stats] of this.statsPerInstance) {
      const recentCount = stats.recentDecodes.filter(
        (t) => t > oneMinuteAgo,
      ).length;
      result[id] = {
        decodesPerMinute: recentCount,
        uniqueCallsigns: stats.uniqueCallsigns.size,
      };
    }

    return result;
  }

  // --------------------------------------------------------------------------
  // Datagram Handling
  // --------------------------------------------------------------------------

  private handleDatagram(buf: Buffer): void {
    if (buf.length < 12) return; // too short for header

    const reader = new QDataStreamReader(buf);

    // Validate magic number
    const magic = reader.readUInt32();
    if (magic !== WSJTX_MAGIC) return;

    // Schema version
    const schema = reader.readUInt32();
    if (schema > WSJTX_SCHEMA_VERSION + 1) return; // allow minor future versions

    // Message type
    const msgType = reader.readUInt32();

    // Instance ID
    const instanceId = reader.readUtf8() ?? "WSJT-X";

    switch (msgType) {
      case WSJTXMessageType.Heartbeat:
        // Heartbeat — no action needed (could track liveness)
        break;

      case WSJTXMessageType.Status:
        this.parseStatus(reader, instanceId);
        break;

      case WSJTXMessageType.Decode:
        this.parseDecode(reader, instanceId);
        break;

      case WSJTXMessageType.Clear:
        this.parseClear(reader, instanceId);
        break;

      case WSJTXMessageType.QSOLogged:
        this.parseQSOLogged(reader, instanceId);
        break;

      case WSJTXMessageType.LoggedADIF:
        this.parseLoggedADIF(reader, instanceId);
        break;

      default:
        // Unknown or unhandled message type — silently ignore
        break;
    }
  }

  // --------------------------------------------------------------------------
  // Message Parsers
  // --------------------------------------------------------------------------

  private parseStatus(reader: QDataStreamReader, instanceId: string): void {
    const dialFrequency = Number(this.readUInt64(reader)); // QDataStream encodes as uint64
    const mode = reader.readUtf8() ?? "";
    const dxCall = reader.readUtf8();
    reader.readUtf8(); // report — skip
    reader.readUtf8(); // tx mode — skip
    const txEnabled = reader.readBool();
    reader.readBool(); // transmitting — skip
    const decoding = reader.readBool();
    const rxDF = reader.readUInt32();
    const txDF = reader.readUInt32();
    reader.readUtf8(); // DE call — skip
    reader.readUtf8(); // DE grid — skip
    const dxGrid = reader.readUtf8();

    const status: WSJTXStatus = {
      frequency: dialFrequency,
      mode,
      dxCall: dxCall ?? undefined,
      dxGrid: dxGrid ?? undefined,
      txEnabled,
      decoding,
      rxDF,
      txDF,
    };

    for (const handler of this.statusHandlers) {
      try {
        handler(status, instanceId);
      } catch {
        /* swallow */
      }
    }
  }

  private parseDecode(reader: QDataStreamReader, instanceId: string): void {
    const isNew = reader.readBool();
    const time = reader.readQTime();
    const snr = reader.readInt32();
    const deltaTime = reader.readDouble();
    const deltaFrequency = reader.readUInt32();
    const mode = reader.readUtf8() ?? "";
    const message = reader.readUtf8() ?? "";
    const lowConfidence = reader.readBool();

    // Extract callsign and grid from the FT8/FT4 message text
    const { callsign, grid } = extractCallInfo(message);

    // Track statistics
    if (!this.statsPerInstance.has(instanceId)) {
      this.statsPerInstance.set(instanceId, createDecodeStats());
    }
    recordDecode(this.statsPerInstance.get(instanceId)!, callsign);

    const decode: WSJTXDecode = {
      isNew,
      time,
      snr,
      deltaTime,
      deltaFrequency,
      mode,
      message,
      lowConfidence,
      callsign,
      grid,
    };

    for (const handler of this.decodeHandlers) {
      try {
        handler(decode, instanceId);
      } catch {
        /* swallow */
      }
    }
  }

  private parseClear(reader: QDataStreamReader, instanceId: string): void {
    // Optional window field (added in later schema versions)
    let window: string | undefined;
    try {
      if (reader.remaining >= 1) {
        const windowId = reader.readUInt8();
        window =
          windowId === 0
            ? "band_activity"
            : windowId === 1
              ? "rx_frequency"
              : undefined;
      }
    } catch {
      // older WSJT-X versions may not include the window field
    }

    for (const handler of this.clearHandlers) {
      try {
        handler(window, instanceId);
      } catch {
        /* swallow */
      }
    }
  }

  private parseQSOLogged(reader: QDataStreamReader, instanceId: string): void {
    this.readQDateTime(reader); // dateTimeOff — skip
    const callsign = reader.readUtf8() ?? "";
    const grid = reader.readUtf8();
    const dialFrequency = Number(this.readUInt64(reader));
    const mode = reader.readUtf8() ?? "";
    const reportSent = reader.readUtf8() ?? "";
    const reportReceived = reader.readUtf8() ?? "";
    const txPower = reader.readUtf8() ?? "";
    const comments = reader.readUtf8() ?? "";
    reader.readUtf8(); // op name — skip
    this.readQDateTime(reader); // dateTimeOn — skip

    const qso: WSJTXQSOLogged = {
      callsign,
      grid: grid ?? undefined,
      frequency: dialFrequency,
      mode,
      reportSent,
      reportReceived,
      txPower,
      comments,
      timestamp: new Date().toISOString(),
    };

    for (const handler of this.qsoLoggedHandlers) {
      try {
        handler(qso, instanceId);
      } catch {
        /* swallow */
      }
    }
  }

  private parseLoggedADIF(reader: QDataStreamReader, instanceId: string): void {
    const adifRecord = reader.readUtf8();

    // We emit this as a QSO logged event with just the ADIF record
    // Parse basic fields from ADIF
    const callsign = extractAdifField(adifRecord, "CALL") ?? "";
    const grid = extractAdifField(adifRecord, "GRIDSQUARE");
    const freqStr = extractAdifField(adifRecord, "FREQ");
    const mode = extractAdifField(adifRecord, "MODE") ?? "";

    const qso: WSJTXQSOLogged = {
      callsign,
      grid: grid ?? undefined,
      frequency: freqStr ? parseFloat(freqStr) * 1_000_000 : 0, // MHz to Hz
      mode,
      reportSent: extractAdifField(adifRecord, "RST_SENT") ?? "",
      reportReceived: extractAdifField(adifRecord, "RST_RCVD") ?? "",
      txPower: extractAdifField(adifRecord, "TX_PWR") ?? "",
      comments: extractAdifField(adifRecord, "COMMENT") ?? "",
      timestamp: new Date().toISOString(),
      adifRecord: adifRecord ?? undefined,
    };

    for (const handler of this.qsoLoggedHandlers) {
      try {
        handler(qso, instanceId);
      } catch {
        /* swallow */
      }
    }
  }

  // --------------------------------------------------------------------------
  // QDataStream Helpers
  // --------------------------------------------------------------------------

  /**
   * Read a uint64 from the stream. JavaScript doesn't have native uint64
   * support, so we read as two uint32 values and combine.
   * Frequencies fit comfortably in Number's safe integer range.
   */
  private readUInt64(reader: QDataStreamReader): number {
    const high = reader.readUInt32();
    const low = reader.readUInt32();
    return high * 0x100000000 + low;
  }

  /**
   * Read a QDateTime from the stream.
   * Format: int64 Julian Day Number + uint32 ms since midnight + uint8 timespec
   */
  private readQDateTime(reader: QDataStreamReader): Date {
    const julianHigh = reader.readInt32();
    const julianLow = reader.readUInt32();
    const julianDay = julianHigh * 0x100000000 + julianLow;
    const msFromMidnight = reader.readUInt32();
    reader.readUInt8(); // timeSpec — skip

    // Convert Julian Day to Unix timestamp
    // Julian Day 2440587.5 = Unix epoch (1970-01-01T00:00:00Z)
    const unixDays = julianDay - 2440588;
    const ms = unixDays * 86400000 + msFromMidnight;
    return new Date(ms);
  }

  // --------------------------------------------------------------------------
  // Event Emission
  // --------------------------------------------------------------------------

  private emitError(error: Error): void {
    for (const handler of this.errorHandlers) {
      try {
        handler(error);
      } catch {
        /* swallow */
      }
    }
  }
}

// ============================================================================
// ADIF Field Extraction Helper
// ============================================================================

function extractAdifField(adif: string | null, field: string): string | null {
  if (!adif) return null;
  const regex = new RegExp(`<${field}:\\d+(?::[^>]*)?>([^<]*)`, "i");
  const match = adif.match(regex);
  return match ? match[1].trim() : null;
}

File: daemon/propulse-integrations/src/wsjtx.rs
use std::time::SystemTime;

use regex::Regex;
use serde::{Deserialize, Serialize};
use tokio::net::UdpSocket;

const WSJTX_MAGIC: u32 = 0xadbccbda;
const WSJTX_SCHEMA_VERSION: u32 = 2;

#[repr(u32)]
enum WSJTXMessageType {
  Heartbeat = 0,
  Status = 1,
  Decode = 2,
  Clear = 3,
  QSOLogged = 5,
  LoggedADIF = 12,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WSJTXStatus {
  pub frequency: u64,
  pub mode: String,
  #[serde(skip_serializing_if = "Option::is_none", rename = "dxCall")]
  pub dx_call: Option<String>,
  #[serde(skip_serializing_if = "Option::is_none", rename = "dxGrid")]
  pub dx_grid: Option<String>,
  #[serde(rename = "txEnabled")]
  pub tx_enabled: bool,
  pub decoding: bool,
  #[serde(rename = "rxDF")]
  pub rx_df: u32,
  #[serde(rename = "txDF")]
  pub tx_df: u32,
  #[serde(skip_serializing_if = "Option::is_none", rename = "instanceId")]
  pub instance_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WSJTXDecode {
  #[serde(rename = "isNew")]
  pub is_new: bool,
  pub time: u32,
  pub snr: i32,
  #[serde(rename = "deltaTime")]
  pub delta_time: f64,
  #[serde(rename = "deltaFrequency")]
  pub delta_frequency: u32,
  pub mode: String,
  pub message: String,
  #[serde(rename = "lowConfidence")]
  pub low_confidence: bool,
  #[serde(skip_serializing_if = "Option::is_none")]
  pub callsign: Option<String>,
  #[serde(skip_serializing_if = "Option::is_none")]
  pub grid: Option<String>,
  #[serde(skip_serializing_if = "Option::is_none", rename = "instanceId")]
  pub instance_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WSJTXQSOLogged {
  pub callsign: String,
  #[serde(skip_serializing_if = "Option::is_none")]
  pub grid: Option<String>,
  pub frequency: u64,
  pub mode: String,
  #[serde(rename = "reportSent")]
  pub report_sent: String,
  #[serde(rename = "reportReceived")]
  pub report_received: String,
  #[serde(rename = "txPower")]
  pub tx_power: String,
  pub comments: String,
  pub timestamp: String,
  #[serde(skip_serializing_if = "Option::is_none", rename = "adifRecord")]
  pub adif_record: Option<String>,
  #[serde(skip_serializing_if = "Option::is_none", rename = "instanceId")]
  pub instance_id: Option<String>,
}

#[derive(Debug, Clone)]
pub enum WSJTXEvent {
  Status(WSJTXStatus),
  Decode(WSJTXDecode),
  QsoLogged(WSJTXQSOLogged),
  Clear { window: Option<String>, instance_id: Option<String> },
}

pub async fn run_wsjtx_listener(
  port: u16,
  mut on_event: impl FnMut(WSJTXEvent) + Send + 'static,
  mut should_stop: impl FnMut() -> bool + Send + 'static,
) -> anyhow::Result<()> {
  let socket = UdpSocket::bind(("127.0.0.1", port)).await?;
  let mut buf = vec![0u8; 64 * 1024];

  loop {
    if should_stop() {
      break;
    }

    let (len, _) = socket.recv_from(&mut buf).await?;
    if len < 12 {
      continue;
    }
    if let Ok(Some(ev)) = parse_datagram(&buf[..len]) {
      on_event(ev);
    }
  }

  Ok(())
}

fn parse_datagram(buf: &[u8]) -> anyhow::Result<Option<WSJTXEvent>> {
  let mut r = QDataStreamReader::new(buf);
  let magic = r.read_u32()?;
  if magic != WSJTX_MAGIC {
    return Ok(None);
  }

  let schema = r.read_u32()?;
  if schema > WSJTX_SCHEMA_VERSION + 1 {
    return Ok(None);
  }

  let msg_type = r.read_u32()?;
  let instance_id = r.read_utf8()?.unwrap_or_else(|| "WSJT-X".to_string());

  match msg_type {
    x if x == WSJTXMessageType::Heartbeat as u32 => Ok(None),
    x if x == WSJTXMessageType::Status as u32 => parse_status(&mut r, instance_id).map(Some),
    x if x == WSJTXMessageType::Decode as u32 => parse_decode(&mut r, instance_id).map(Some),
    x if x == WSJTXMessageType::Clear as u32 => parse_clear(&mut r, instance_id).map(Some),
    x if x == WSJTXMessageType::QSOLogged as u32 => parse_qso_logged(&mut r, instance_id).map(Some),
    x if x == WSJTXMessageType::LoggedADIF as u32 => parse_logged_adif(&mut r, instance_id).map(Some),
    _ => Ok(None),
  }
}

fn parse_status(r: &mut QDataStreamReader<'_>, instance_id: String) -> anyhow::Result<WSJTXEvent> {
  let dial_frequency = r.read_u64_as_number()?;
  let mode = r.read_utf8()?.unwrap_or_default();
  let dx_call = r.read_utf8()?;
  let _report = r.read_utf8()?;
  let _tx_mode = r.read_utf8()?;
  let tx_enabled = r.read_bool()?;
  let _transmitting = r.read_bool()?;
  let decoding = r.read_bool()?;
  let rx_df = r.read_u32()?;
  let tx_df = r.read_u32()?;
  let _de_call = r.read_utf8()?;
  let _de_grid = r.read_utf8()?;
  let dx_grid = r.read_utf8()?;

  Ok(WSJTXEvent::Status(WSJTXStatus {
    frequency: dial_frequency,
    mode,
    dx_call,
    dx_grid,
    tx_enabled,
    decoding,
    rx_df,
    tx_df,
    instance_id: Some(instance_id),
  }))
}

fn parse_decode(r: &mut QDataStreamReader<'_>, instance_id: String) -> anyhow::Result<WSJTXEvent> {
  let is_new = r.read_bool()?;
  let time = r.read_u32()?; // QTime ms since midnight
  let snr = r.read_i32()?;
  let delta_time = r.read_f64()?;
  let delta_frequency = r.read_u32()?;
  let mode = r.read_utf8()?.unwrap_or_default();
  let message = r.read_utf8()?.unwrap_or_default();
  let low_confidence = r.read_bool()?;

  let (callsign, grid) = extract_call_info(&message);

  Ok(WSJTXEvent::Decode(WSJTXDecode {
    is_new,
    time,
    snr,
    delta_time,
    delta_frequency,
    mode,
    message,
    low_confidence,
    callsign,
    grid,
    instance_id: Some(instance_id),
  }))
}

fn parse_clear(r: &mut QDataStreamReader<'_>, instance_id: String) -> anyhow::Result<WSJTXEvent> {
  let window = if r.remaining() >= 1 {
    match r.read_u8()? {
      0 => Some("band_activity".to_string()),
      1 => Some("rx_frequency".to_string()),
      _ => None,
    }
  } else {
    None
  };

  Ok(WSJTXEvent::Clear {
    window,
    instance_id: Some(instance_id),
  })
}

fn parse_qso_logged(r: &mut QDataStreamReader<'_>, instance_id: String) -> anyhow::Result<WSJTXEvent> {
  let _date_time_off = r.read_qdatetime()?;
  let callsign = r.read_utf8()?.unwrap_or_default();
  let grid = r.read_utf8()?;
  let dial_frequency = r.read_u64_as_number()?;
  let mode = r.read_utf8()?.unwrap_or_default();
  let report_sent = r.read_utf8()?.unwrap_or_default();
  let report_received = r.read_utf8()?.unwrap_or_default();
  let tx_power = r.read_utf8()?.unwrap_or_default();
  let comments = r.read_utf8()?.unwrap_or_default();
  let _op = r.read_utf8()?;
  let _date_time_on = r.read_qdatetime()?;

  Ok(WSJTXEvent::QsoLogged(WSJTXQSOLogged {
    callsign,
    grid,
    frequency: dial_frequency,
    mode,
    report_sent,
    report_received,
    tx_power,
    comments,
    timestamp: chrono::Utc::now().to_rfc3339(),
    adif_record: None,
    instance_id: Some(instance_id),
  }))
}

fn parse_logged_adif(r: &mut QDataStreamReader<'_>, instance_id: String) -> anyhow::Result<WSJTXEvent> {
  let adif = r.read_utf8()?;
  let callsign = extract_adif_field(adif.as_deref(), "CALL").unwrap_or_default();
  let grid = extract_adif_field(adif.as_deref(), "GRIDSQUARE");
  let freq_str = extract_adif_field(adif.as_deref(), "FREQ");
  let mode = extract_adif_field(adif.as_deref(), "MODE").unwrap_or_default();

  let frequency = freq_str
    .and_then(|s| s.parse::<f64>().ok())
    .map(|mhz| (mhz * 1_000_000.0) as u64)
    .unwrap_or(0);

  Ok(WSJTXEvent::QsoLogged(WSJTXQSOLogged {
    callsign,
    grid,
    frequency,
    mode,
    report_sent: extract_adif_field(adif.as_deref(), "RST_SENT").unwrap_or_default(),
    report_received: extract_adif_field(adif.as_deref(), "RST_RCVD").unwrap_or_default(),
    tx_power: extract_adif_field(adif.as_deref(), "TX_PWR").unwrap_or_default(),
    comments: extract_adif_field(adif.as_deref(), "COMMENT").unwrap_or_default(),
    timestamp: chrono::Utc::now().to_rfc3339(),
    adif_record: adif,
    instance_id: Some(instance_id),
  }))
}

// ────────────────────────────────────────────────────────────────────────────
// QDataStream reader helpers
// ────────────────────────────────────────────────────────────────────────────

struct QDataStreamReader<'a> {
  buf: &'a [u8],
  off: usize,
}

impl<'a> QDataStreamReader<'a> {
  fn new(buf: &'a [u8]) -> Self {
    Self { buf, off: 0 }
  }

  fn remaining(&self) -> usize {
    self.buf.len().saturating_sub(self.off)
  }

  fn read_u8(&mut self) -> anyhow::Result<u8> {
    if self.remaining() < 1 {
      return Err(anyhow::anyhow!("buffer underflow u8"));
    }
    let v = self.buf[self.off];
    self.off += 1;
    Ok(v)
  }

  fn read_bool(&mut self) -> anyhow::Result<bool> {
    Ok(self.read_u8()? != 0)
  }

  fn read_u32(&mut self) -> anyhow::Result<u32> {
    if self.remaining() < 4 {
      return Err(anyhow::anyhow!("buffer underflow u32"));
    }
    let v = u32::from_be_bytes(self.buf[self.off..self.off + 4].try_into()?);
    self.off += 4;
    Ok(v)
  }

  fn read_i32(&mut self) -> anyhow::Result<i32> {
    if self.remaining() < 4 {
      return Err(anyhow::anyhow!("buffer underflow i32"));
    }
    let v = i32::from_be_bytes(self.buf[self.off..self.off + 4].try_into()?);
    self.off += 4;
    Ok(v)
  }

  fn read_f64(&mut self) -> anyhow::Result<f64> {
    if self.remaining() < 8 {
      return Err(anyhow::anyhow!("buffer underflow f64"));
    }
    let v = f64::from_be_bytes(self.buf[self.off..self.off + 8].try_into()?);
    self.off += 8;
    Ok(v)
  }

  fn read_utf8(&mut self) -> anyhow::Result<Option<String>> {
    let len = self.read_u32()?;
    if len == 0xffff_ffff {
      return Ok(None);
    }
    let len = len as usize;
    if self.remaining() < len {
      return Err(anyhow::anyhow!("buffer underflow utf8"));
    }
    let s = std::str::from_utf8(&self.buf[self.off..self.off + len])?.to_string();
    self.off += len;
    Ok(Some(s))
  }

  fn read_u64_as_number(&mut self) -> anyhow::Result<u64> {
    // QDataStream encodes u64 as two u32 big-endian parts.
    let high = self.read_u32()? as u64;
    let low = self.read_u32()? as u64;
    Ok((high << 32) | low)
  }

  fn read_qdatetime(&mut self) -> anyhow::Result<SystemTime> {
    // Format: int64 Julian day + u32 ms since midnight + u8 timespec
    let julian_high = self.read_i32()? as i64;
    let julian_low = self.read_u32()? as i64;
    let julian = (julian_high << 32) | (julian_low & 0xffff_ffff);
    let ms_midnight = self.read_u32()? as i64;
    let _timespec = self.read_u8()?;

    // Julian Day 2440587.5 = Unix epoch. In WSJT-X wire format, this matches the
    // approach used in the existing bridge implementation: subtract 2440588.
    let unix_days = julian - 2_440_588;
    let ms = unix_days * 86_400_000 + ms_midnight;
    Ok(SystemTime::UNIX_EPOCH + std::time::Duration::from_millis(ms.max(0) as u64))
  }
}

// ────────────────────────────────────────────────────────────────────────────
// Message parsing helpers
// ────────────────────────────────────────────────────────────────────────────

fn extract_call_info(message: &str) -> (Option<String>, Option<String>) {
  // CQ [dir] CALL GRID
  let cq_re = Regex::new(r"^CQ\s+(?:(\w{2,4})\s+)?([A-Z0-9/]{3,})\s+([A-R]{2}\d{2}(?:[a-x]{2})?)$").ok();
  if let Some(re) = cq_re {
    if let Some(caps) = re.captures(message.trim()) {
      return (
        caps.get(2).map(|m| m.as_str().to_string()),
        caps.get(3).map(|m| m.as_str().to_string()),
      );
    }
  }

  let grid_re = Regex::new(r"^([A-Z0-9/]{3,})\s+([A-Z0-9/]{3,})\s+([A-R]{2}\d{2}(?:[a-x]{2})?)$").ok();
  if let Some(re) = grid_re {
    if let Some(caps) = re.captures(message.trim()) {
      return (
        caps.get(2).map(|m| m.as_str().to_string()),
        caps.get(3).map(|m| m.as_str().to_string()),
      );
    }
  }

  let report_re = Regex::new(r"^([A-Z0-9/]{3,})\s+([A-Z0-9/]{3,})\s+(R?[+-]\d{2}|RR73|RRR|73)$").ok();
  if let Some(re) = report_re {
    if let Some(caps) = re.captures(message.trim()) {
      return (caps.get(2).map(|m| m.as_str().to_string()), None);
    }
  }

  (None, None)
}

fn extract_adif_field(adif: Option<&str>, field: &str) -> Option<String> {
  let adif = adif?;
  let re = Regex::new(&format!(r"(?i)<{field}:\d+(?::[^>]*)?>([^<]*)")).ok()?;
  let caps = re.captures(adif)?;
  Some(caps.get(1)?.as_str().trim().to_string())
}
